feat http-client: add curl easy wrapper prototype

shindaq opened the pull request.

Commit:
feat http-client: add curl easy wrapper prototype

Commit:
fix

Commit:
fix

Commit:
add gitignore

Commit:
fix

Commit:
fix: fixed CurlPPEasy::SetHeaders leads to memory leaks

uwa9k073 commented on include/client/client.hpp:
```diff
@@ -0,0 +1,15 @@
+  Client();
+
+private:
+  std::unique_ptr<CURL, decltype(&curl_easy_cleanup)> _curl;
```
почему unique_ptr?

shindaq replied:
а я не ебу как с шаредом сделать чтобы работало, я в целом с умными указателями никогда не работал

uwa9k073 commented on include/curlpp/curlpp_exceptions.hpp:
```diff
@@ -0,0 +1,24 @@
+
+enum class CurlExceptionTypes { kCurlInitFailure, kCurlSlistAppendFailure };
+
+class CurlException final : public std::exception {
```
почему бы не сделать два разных исключения?

shindaq replied:
ну можно и так в целом, но мне было впадлу совершенно немного

uwa9k073 replied:
а теперь не впадлу

uwa9k073 commented on include/curlpp/response.hpp:
```diff
@@ -0,0 +1,5 @@
+
+namespace curlpp {
+class Response {};
+} // namespace curlpp
```
зачем тебе пустые `Request` и `Response`?

shindaq replied:
потому что они еще не написаны были

Commit:
feat http-client: removed CurlPPEasy::SetHeaders with CurlPPEasy::SetSlistOption
SetSlistOption no longer linked to http headers but any libcurl options which depends on curl_slist

uwa9k073 commented on include/curlpp/curl_easy.cpp:
```diff
@@ -10,69 +10,52 @@ namespace curlpp {
-  _slist.reset(slist);
+  std::unique_ptr<curl_slist, decltype(&curl_slist_free_all)> slist(
+      raw_slist, curl_slist_free_all);
+  _slists.push_back(std::move(slist));
```
ебанный в рот

uwa9k073 commented on include/curlpp/curl_easy.hpp:
```diff
@@ -11,30 +11,25 @@ class CurlPPEasy final {
 private:
   std::unique_ptr<CURL, decltype(&curl_easy_cleanup)> _curl;
-  std::unique_ptr<curl_slist, decltype(&curl_slist_free_all)> _slist;
+  std::vector<std::unique_ptr<curl_slist, decltype(&curl_slist_free_all)>>
```
`using SomeType = dectype(&curl_slist_free_all)`

uwa9k073 commented on include/curlpp/curl_easy.cpp:
```diff
@@ -10,69 +10,52 @@ namespace curlpp {
   }
-  curl_easy_setopt(_curl.get(), CURLOPT_HTTPHEADER, slist);
-  _slist.reset(slist);
+  std::unique_ptr<curl_slist, decltype(&curl_slist_free_all)> slist(
```
`using SList = std::unique_ptr<curl_slist, decltype(&curl_slist_free_all)>`

uwa9k073 commented on include/client/client.hpp:
```diff
@@ -0,0 +1,15 @@
+  Client();
+
+private:
+  std::unique_ptr<CURL, decltype(&curl_easy_cleanup)> _curl;
```
сюда тоже аляс нужен

Commit:
fix http-client: removed long unique_ptr declarations with aliases
